### DISCUSSION TITLE
Discovering the weakness of the RealPLKSR model architecture (yet to be confirmed).

### DISCUSSION THREAD

**Discussion by avan06** [Phhofm/models#8] [Q&A] [answered]
Hi Philip,

First of all, thank you for training so many excellent models. Recently, I’ve been using the 4xNomosWebPhoto_RealPLKSR model in my program. However, I’ve found that when testing with the sample images provided on your page, I couldn’t achieve the expected results no matter what I tried. Later, I identified the cause and wanted to share this case with you.

In my program, when an input image’s height is less than 300, the program doubles its size. When I tested with the smaller sample images, doubling their size first and then using the 4xNomosWebPhoto_RealPLKSR model to upscale them 4x, the output was far from satisfactory. However, when I removed the resizing step, the output was as expected.

___

Below are the case images:

Original image1
![b01](https://github.com/user-attachments/assets/c9699eb6-0ff8-40c7-80ab-77bdbb4d7108)

Model output from the original image1
![b01_test2](https://github.com/user-attachments/assets/60f22a49-2398-4a7b-b83a-37280fd80f03)

Original image1 resized to double the size
![b01](https://github.com/user-attachments/assets/e3f7fd2f-b253-4464-a2ca-5dd3f2e9a34a)

Model output from the resized image1
![b01_test1](https://github.com/user-attachments/assets/186ae9a6-16a6-485e-ab3b-dfad13a1b3cb)

___

Original image2
![b03](https://github.com/user-attachments/assets/ca33a62b-5c07-412e-a6fc-acea9096aad5)

Model output from the original image2
![b03_test2](https://github.com/user-attachments/assets/120fcc75-855f-4915-b1f1-927be48b8c78)

Original image2 resized to double the size
![b03](https://github.com/user-attachments/assets/f9c57251-8e70-4935-b185-296ddd39d567)

Model output from the resized image2
![b03_test1](https://github.com/user-attachments/assets/d9d34f8b-aa07-4b73-ba82-64d59110d66f)

___

Original image3
![b06](https://github.com/user-attachments/assets/5f1d3f97-6919-4d4e-ac38-09b28cd2f5f2)

Model output from the original image3
![b06_test2](https://github.com/user-attachments/assets/5d5ac93d-551f-49ac-93f5-3013f448905b)

Original image3 resized to double the size
![b06](https://github.com/user-attachments/assets/0e7492be-2dfb-4734-87e4-6d560fcba0ad)

Model output from the resized image3
![b06_test1](https://github.com/user-attachments/assets/b14c27e5-1717-47e6-bf42-98e39c42b06e)

___

Original image4
![b07](https://github.com/user-attachments/assets/2694a8d9-9ade-4aad-8d2b-60dd4c353b01)

Model output from the original image4
![b07_test2](https://github.com/user-attachments/assets/f79885f8-9cdb-408c-a48c-ad60778d3086)

Original image4 resized to double the size
![b07](https://github.com/user-attachments/assets/9d6ebf2e-cab1-4fc5-86eb-212860329cf6)

Model output from the resized image4
![b07_test1](https://github.com/user-attachments/assets/19d9740b-1b9b-4c10-9fe5-9658a3bdc989)

___

Original image5
![b09](https://github.com/user-attachments/assets/8234f293-4832-4470-98e2-38d9f54dff5c)

Model output from the original image5
![b09_test2](https://github.com/user-attachments/assets/0b922cd4-0ec2-4d83-b237-69bd035e3fc9)

Original image5 resized to double the size
![b09](https://github.com/user-attachments/assets/3772fe3a-8fd9-4b99-bf2d-6b6805b9f9e2)

Model output from the resized image5
![b09_test1](https://github.com/user-attachments/assets/8ff4919f-f2cf-46eb-81b0-85b66fef6007)

___

As shown in the test results above, the RealPLKSR model produces suboptimal results in this specific scenario. I’m not sure what the cause might be, but thank you for looking into it!

**Comment by Phhofm** [comment, marked as answer]
tldr: model is not trained for that use case. Its not an architecture thing but degradation/training thing.


Hey, thanks for the post
Yeah this has a very simple reason, that model was not trained for upsized content.
Well its not entirely true, the training (or rather, LR creation) setting was this
![image](https://github.com/user-attachments/assets/74e6bbca-b1e1-4374-8d16-83b64b12083f)
Meaning it might be able to handle 1.25x upsized content. 
I simply didnt think someone would 2x size content before the upscale since if an image (like a photo) gets put on a website, to increase page speed, it often gets sized down and compressed (not the other way around).

So to answer your question, this model was simply not trained to handle someone preprocessing input with a 2x resize. Thats also what you are seeing in your tests.

Even though this model was trained to be used without this purpose in mind, I can show something that would improve your results, by simply changing your order (if you need those image dimensions):
First we resize 2x with cubic, then upscale with this model, the expected bad result
![2Cubic_nomoswebphoto](https://github.com/user-attachments/assets/f3e3e9b6-0cd6-4343-9eba-8ee1ff618bfd)
Now we instead upscale first, then resize 2x with cubic, it should look better
![upscale_resize_cubic](https://github.com/user-attachments/assets/12c8138b-bd0b-42b9-aa9e-45f07c1f5883)
We could alternatively also upscale, then upscale that output again, then resize 0.5x with cubic
![up_up_cubicdown](https://github.com/user-attachments/assets/0b3fb40d-ae3e-47c6-a1e4-ea2d7d7f5953)

And just to show in general, an otf trained model might be better equipped to handle a specific case where someone 2x sized content before upscaling, because it has a chance to up_up with up to 1.5x and then 1.2x in the two degradation steps which in theory could come up to 1.8x input during training
2x cubic sized, then upscaled, with 4xBHI_realplksr_dysample_otf_nn (so still a realplksr model because I wanted to show the training difference, not some architecture difference)
![2Cubic_bhi_otf_nn](https://github.com/user-attachments/assets/fdd6aef0-8099-4214-b3a3-e22d0e2635b5)

Hope this helps for understanding

**Comment by avan06** [comment]
Thank you very much to the model author for the detailed explanation and for suggesting improvement methods. I now fully understand!  

Regarding the 2x enlargement of small images, I only noticed it this time because I duplicated someone else's Image Upscale Space and modified it to support the RealPLKSR model. I’ve already removed the resizing code, so there’s no issue now.  

Once again, thank you so much!
